pages/api/update-member.tsx

```diff
@@ -0,0 +1,163 @@
+    const referencesToAdd: DocumentReference[] = newReferences.filter((ref) => !oldReferences.includes(ref));
+    const memberRefPublic = await getMemberRefPublic(uid);
+    await deleteReferencesPublic(memberRefPublic, referencesToDelete);
+    await addMemberToReferencesPublic(memberRefPublic, referencesToAdd, currentUser);
+
+    // Need to convert to admin.firestore.DocumentReference, difference in how they're handled
+    const adminReferencesToAdd = referencesToAdd.map((ref) => admin.firestore().doc(ref.path));
+    const adminReferencesToDelete = referencesToDelete.map((ref) => admin.firestore().doc(ref.path));
+    const oldData = doc.get(fieldName);
+    let writeResult = null;
+    if (adminReferencesToDelete.length !== 0) {
+      let writeResult = await docRef.update({
+        [fieldName]: admin.firestore.FieldValue.arrayRemove(...adminReferencesToDelete),
+        last_modified: admin.firestore.FieldValue.serverTimestamp(),
+        last_modified_by: currentUser || "admin edit",
+      });
+      console.log(`Deleted references ${adminReferencesToDelete} from ${fieldName} for ${uid}: ${writeResult}`);
+    }
+    if (adminReferencesToAdd.length !== 0) {
+      let writeResult = await docRef.update({
+        [fieldName]: admin.firestore.FieldValue.arrayUnion(...adminReferencesToAdd),
+        last_modified: admin.firestore.FieldValue.serverTimestamp(),
+        last_modified_by: currentUser || "admin edit",
+      });
+      console.log(`Added references ${adminReferencesToAdd} to ${fieldName} for ${uid}: ${writeResult}`);
+    }
+    console.log(`Updated ${fieldName} from ${oldData} to ${newData} for ${uid}`);
+    return writeResult;
+  } else if (suggestedFilter && arrayFields.includes(fieldName)) {
+    for (const newFitler of newData) {
+      if (typeof newFitler !== "string") {
+        throw new Error(`Field ${fieldName} is supposed to be a string array, but newData contains a non-string`);
+      }
+      const memberRefPublic = await getMemberRefPublic(uid);
+      const newLabelRef = await addLabelRefPublic(newFitler, fieldNameToTable[fieldName]);
+      await addMemberToLabelsPublic([newLabelRef], memberRefPublic);
+      const writeResult = await docRef.update({
+        [fieldName]: admin.firestore.FieldValue.arrayUnion(...[admin.firestore().doc(newLabelRef.path)]),
+        last_modified: admin.firestore.FieldValue.serverTimestamp(),
+        last_modified_by: currentUser || "admin edit",
+      });
+      console.log(writeResult);
+    }
+  } else {
+    const oldData = doc.get(fieldName);
+    const writeResult = await docRef.update({
+      [fieldName]: newData,
+      last_modified: admin.firestore.FieldValue.serverTimestamp(),
+      last_modified_by: currentUser || "admin edit",
+    });
+    console.log(`Updated ${fieldName} from ${oldData} to ${newData} for ${uid}`);
+    return writeResult;
+  }
+};
+
+export default async function handler(req, res) {
+  if (req.method !== "PUT") {
```

Comment:
I entirely agree on this and I appreciate this feedback.

In terms of the `PUT` vs `PATCH`, I wasn't aware of the idempotence distinguishment, that best practice makes sense. Also since we're handling most if not all of the fields to update in a member document, it makes sense to transition to more scalable method using PUT.

I'll just have to think a bit carefully on how it's implemented but I will definitely be refactoring this

Reply:
refactored in the next commit